Source/santametricservice/Formats/SNTMetricRawJSONFormat.m

```diff
@@ -0,0 +1,100 @@
+/// Copyright 2021 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+#import "Source/common/SNTLogging.h"
+
+#import "Source/santametricservice/Formats/SNTMetricRawJSONFormat.h"
+
+@implementation SNTMetricRawJSONFormat {
+  NSDateFormatter *_dateFormatter;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _dateFormatter = [[NSDateFormatter alloc] init];
+    [_dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];
+  }
+  return self;
+}
+
+- (NSArray *)normalizeArray:(NSArray *)arr {
+  NSMutableArray *normalized = [NSMutableArray arrayWithArray:arr];
+
+  for (int i = 0; i < [arr count]; i++) {
+    if ([arr[i] isKindOfClass:[NSArray class]]) {
+      normalized[i] = [self normalizeArray:(NSArray *)arr[i]];
+    } else if ([arr[i] isKindOfClass:[NSDictionary class]]) {
+      normalized[i] = [self normalize:(NSDictionary *)arr[i]];
+    }
+  }
+
+  return normalized;
+}
+
+/**
+ * Normalizes the metrics dictionary for exporting to JSON
+ **/
+- (NSDictionary *)normalize:(NSDictionary *)metrics {
+  // Convert NSDate's to RFC3339 in strings as NSDate's cannot be serialized
+  // to JSON.
+  NSMutableDictionary *normalizedMetrics = [NSMutableDictionary dictionaryWithDictionary:metrics];
+
+  for (NSString *key in metrics) {
```

Comment:
For enumerating dictionaries there is a more performant option that avoids having to do the retrieval:

```
[metrics enumerateKeysAndObjectsUsingBlock:^(id key, id value, BOOL* stop) {
  // work with key/value
}];
```

It's arguably less readable but can be worth it if the dictionary has lots of keys.

Reply:
I'd forgotten about that. I've switch over to Block iteration for this and the normalize array function.